File: backend/api/sources.py
"""Лента уровня 1: карточки подключённых источников."""
from fastapi import APIRouter, Depends
from sqlalchemy import select
from sqlalchemy.orm import Session

from backend.db import get_db
from backend.models import Connection, Message, User
from backend.queries import level_distribution
from backend.schemas import SourceCard
from backend.security import current_user
from backend.serializers import message_brief

router = APIRouter(prefix="/api", tags=["sources"])


@router.get("/sources", response_model=list[SourceCard])
def list_sources(
    user: User = Depends(current_user),
    db: Session = Depends(get_db),
) -> list[SourceCard]:
    """Отключённые источники в ленте не показываются вообще (решение №22)."""
    connections = db.execute(
        select(Connection)
        .where(Connection.user_id == user.id, Connection.state.in_(("active", "reauth")))
        .order_by(Connection.kind)
    ).scalars().all()

    cards: list[SourceCard] = []
    for connection in connections:
        stmt = select(Message).where(Message.connection_id == connection.id)
        messages = db.execute(
            stmt.order_by(Message.received_at.desc(), Message.id.desc())
        ).scalars().all()
        urgent = next((m for m in messages if m.effective_level == "CRITICAL"), None) or next(
            (m for m in messages if m.effective_level == "HIGH"), None
        )
        cards.append(
            SourceCard(
                kind=connection.kind,
                state=connection.state,
                account=connection.account,
                last_sync_at=connection.last_sync_at,
                total=len(messages),
                unread=sum(1 for m in messages if not m.is_read),
                distribution=level_distribution(db, stmt),
                urgent=message_brief(urgent) if urgent else None,
            )
        )
    return cards

File: backend/internal/routers/sources.go
package routers

import (
	"net/http"

	"personalinbox/internal/schemas"
	"personalinbox/internal/sqlite"
)

// handleListSources — лента уровня 1. Отключённые источники не показываются
// вообще (решение №22).
func (s *Server) handleListSources(w http.ResponseWriter, r *http.Request) {
	user, ok := s.currentUser(w, r)
	if !ok {
		return
	}
	connections, err := s.db.VisibleConnectionsOf(user.ID)
	if err != nil {
		fail(w, http.StatusInternalServerError, "Не удалось получить источники")
		return
	}

	cards := make([]schemas.SourceCard, 0, len(connections))
	for _, connection := range connections {
		messages, err := s.db.MessagesOfConnection(connection.ID)
		if err != nil {
			fail(w, http.StatusInternalServerError, "Не удалось получить сообщения")
			return
		}
		unread := 0
		for _, message := range messages {
			if !message.IsRead {
				unread++
			}
		}
		cards = append(cards, schemas.SourceCard{
			Kind:         connection.Kind,
			State:        connection.State,
			Account:      connection.Account,
			LastSyncAt:   schemas.TimePtr(connection.LastSyncAt),
			Total:        len(messages),
			Unread:       unread,
			Distribution: sqlite.Distribution(messages),
			Urgent:       urgent(messages),
		})
	}
	respond(w, http.StatusOK, cards)
}

// urgent — самое срочное сообщение источника: сначала CRITICAL, потом HIGH.
func urgent(messages []*sqlite.Message) *schemas.MessageBrief {
	for _, level := range []string{"CRITICAL", "HIGH"} {
		for _, message := range messages {
			if message.EffectiveLevel() == level {
				brief := schemas.MessageBriefOut(message)
				return &brief
			}
		}
	}
	return nil
}
